File: CMR2005CG.vb
Imports System.Text.RegularExpressions

Module CMR2005CG

  Function Quotient(n As Long, m As Long) As Long
    Return If(n / m > 0, Math.Floor(n / m), Math.Ceiling(n / m))
  End Function

  Function Remainder(n As Long, m As Long) As Long
    Return n - Quotient(n, m) * m
  End Function

  ReadOnly Cheats As String() = {"All tracks code", "4WD cars code", "2WD cars code", "Super 2WD cars code", "RWD cars code", "4x4 cars code", "Classic cars code", "Special cars code", "Group B cars code", "Mirror mode code (inaccessible)"}

  Function IsValidCode(text As String) As Boolean
    If Regex.IsMatch(text, "[^0-9]") Then Return False
    Try
      Dim n As Long = CType(text, Long)
      Return True
    Catch
      Return False
    End Try
  End Function

  Sub Main(Args As String())
    Console.WriteLine("Colin McRae Rally 2005 Cheat Generator")
    If Args.Length <> 4 OrElse (Not Args.Contains("-p") AndAlso Not Args.Contains("-P")) OrElse (Not Args.Contains("-c") AndAlso Not Args.Contains("-C")) Then
      Console.WriteLine("Usage: CMR2005CG [commands]")
      Console.WriteLine("Commands  Parameters            Explanation")
      Console.WriteLine("-p        [pc, ps2, psp, xbox]  Set game platform")
      Console.WriteLine("-c        [1, ..., 99999]       Set access code")
      Return
    End If
    Dim ip As Int32 = If(Args.Contains("-p"), Array.IndexOf(Args, "-p"), Array.IndexOf(Args, "-P"))
    Dim ic As Int32 = If(Args.Contains("-c"), Array.IndexOf(Args, "-c"), Array.IndexOf(Args, "-C"))
    If Not IsValidCode(Args(ic + 1)) Then
      Console.WriteLine("Invalid access code")
      Return
    End If
    Dim Platform As String = Args(ip + 1).ToLower()
    Dim AccessCode As Long = CType(Args(ic + 1), Long)
    If AccessCode < 1L Or AccessCode > 99999L Then
      Console.WriteLine("Invalid access code. Valid access codes are in range 1 - 99999")
      Return
    End If
    Dim Array1 As Long()
    Dim Array2 As Long()
    Select Case Platform
      Case "pc"
        Array1 = {331L, 1259L, 241L, 109L, 521L, 853L, 71L, 719L, 941L, 269L}
        Array2 = {72481L, 180307L, 130241L, 392827L, 421019L, 49147L, 32801L, 1296649L, 91249L, 639679L}
      Case "ps2", "xbox"
        Array1 = {859L, 773L, 151L, 47L, 487L, 211L, 617L, 131L, 947L, 313L}
        Array2 = {69119L, 67783L, 70271L, 77929L, 238099L, 148151L, 472751L, 818963L, 1195489L, 839381L}
      Case "psp"
        Array1 = {743L, 1663L, 227L, 991L, 443L, 89L, 571L, 199L, 1373L, 601L}
        Array2 = {35491L, 783019L, 1116491L, 591319L, 194591L, 37369L, 822839L, 86083L, 354661L, 99809L}
      Case Else
        Console.WriteLine("Invalid platform. Enter a platform among pc, ps2, psp, xbox")
        Return
    End Select
    Console.WriteLine("Platform: " + Platform.ToUpper())
    Console.WriteLine("Access code: " + AccessCode.ToString())
    For i As Int32 = 0 To Cheats.Length - 1
      Console.WriteLine(Cheats(i) + ": " + GenerateCode(Array1, Array2, AccessCode, i))
    Next i
  End Sub

  Function GenerateCode(Array1 As Long(), Array2 As Long(), AccessCode As Long, CheatID As Long) As String
    Dim CheatIDMagic As Long = &H13CB5BL * CheatID Mod &H26DDL
    Dim AccessCodeMagic As Long = (AccessCode Mod &H3E8L) Xor (&H20B9L * CheatIDMagic)
    Dim Seed1 As Long = CalculateSeed(AccessCodeMagic Mod &H9ADL)
    Dim Seed2 As Long = CalculateSeed(Remainder(ToSigned32((AccessCodeMagic Xor &H114CF1L) * ((&H41BL * CheatIDMagic) Xor Remainder(Quotient(AccessCode, &H3E8L), &H3E8L))), &H91DL))
    Dim Buffer As Long() = {0L, 0L, 0L, 0L, 0L, 0L}
    Buffer(0) = Remainder(Seed1, 26L) + 65L
    Buffer(1) = Remainder(Quotient(Seed1, 676L), 26L) + 65L
    Buffer(2) = Remainder(Quotient(Seed1, 26L), 26L) + 65L
    Buffer(3) = Remainder(Quotient(Seed2, 26L), 26L) + 65L
    Buffer(4) = Remainder(Quotient(Seed2, 676L), 26L) + 65L
    Buffer(5) = Remainder(Seed2, 26L) + 65L
    Dim BufMidXor As Long = CalculateFeedback(Buffer)
    Dim Feedback1 As Long = ToSigned32((Buffer(0) << 24L) + (Buffer(1) << 16L) + (Buffer(2) << 8L) + Buffer(3))
    Dim Feedback2 As Long = ToSigned32((Buffer(4) << 24L) + (Buffer(5) << 16L) + ((BufMidXor + Remainder(CheatIDMagic Xor &H197ABD9L, Seed1 And &HFFFFFFFFL)) << 8L) + BufMidXor + Remainder(CheatIDMagic Xor &H13478FDDL, Seed2 And &HFFFFFFFFL))
    Dim TmpFeedback2 As Long = 0L
    For i As Int32 = 0 To 41
      TmpFeedback2 = Feedback2
      Feedback2 = Feedback1 Xor Array1(i Mod 10)
      Feedback1 = TmpFeedback2 Xor Feedback1
    Next i
    For i As Int32 = 0 To 276
      TmpFeedback2 = Feedback2
      Feedback2 = Feedback1 Xor Array2(i Mod 10)
      Feedback1 = Feedback1 Xor TmpFeedback2
    Next i
    Buffer(0) = Remainder((Feedback2 >> 24L) And &HFFL, 26L) + 65L
    Buffer(1) = Remainder((Feedback2 >> 16L) And &HFFL, 26L) + 65L
    Buffer(2) = Remainder((Feedback1 >> 24L) And &HFFL, 26L) + 65L
    Buffer(3) = Remainder((Feedback1 >> 16L) And &HFFL, 26L) + 65L
    Buffer(4) = Remainder((Feedback1 >> 8L) And &HFFL, 26L) + 65L
    Buffer(5) = Remainder(Feedback1 And &HFFL, 26L) + 65L
    Dim code As String = ""
    For i As Int32 = 0 To 5
      code += Chr(Buffer(i))
    Next i
    Return code
  End Function

  Function CalculateFeedback(Buffer As Long()) As Long
    Dim Result As Long = 0
    For i As Int32 = 0 To Buffer.Length - 2
      Result += Buffer(i) Xor &H13C501L
    Next i
    Return ToSigned32(Result)
  End Function

  Function ToSigned32(n As Long) As Long
    Return ((n And &HFFFFFFFFL) Xor &H80000000L) - &H80000000L
  End Function

  Function CalculateSeed(n As Long) As Long
    Dim Seed As Long = 1L
    If n <> 0L Then
      Seed = &HF82DL
      For i As Int32 = 0 To n - 2
        Seed = Remainder(ToSigned32(&HF82DL * Seed), &H5243L)
      Next i
    End If
    Return Seed
  End Function

End Module

File: CMR2005CG.java
    import java.util.Arrays;
    
    public class Main
    {

        public static String[] cheats = { "All tracks code", "4WD cars code", "2WD cars code", "Super 2WD cars code", "RWD cars code", "4x4 cars code", "Classic cars code", "Special cars code", "Group B cars code", "Mirror mode code (inaccessible)" };

        private static boolean Contains(String[] array, String text)
        {
            return Arrays.asList(array).indexOf(text) != -1;
        }

        private static boolean IsValidCode(String text)
        {
            try
            {
                long n = Long.parseLong(text, 10);
                return true;
            }
            catch (Exception ex)
            {
                return false;
            }
        }

        public static void main(String[] args)
        {
            System.out.println("Colin McRae Rally 2005 Cheat Generator");
            if (args.length != 4 || (!Contains(args, "-p") && !Contains(args, "-P")) || (!Contains(args, "-c") && !Contains(args, "-C")))
            {
                System.out.println("Usage: CMR2005CG [commands]");
                System.out.println("Commands  Parameters            Explanation");
                System.out.println("-p        [pc, ps2, psp, xbox]  Set game platform");
                System.out.println("-c        [1, ..., 99999]       Set access code");
                return;
            }
            int ip = Contains(args, "-p") ? Arrays.asList(args).indexOf("-p") : Arrays.asList(args).indexOf("-P");
            int ic = Contains(args, "-c") ? Arrays.asList(args).indexOf("-c") : Arrays.asList(args).indexOf("-C");
            if (!IsValidCode(args[ic + 1]))
            {
                System.out.println("Invalid access code");
                return;
            }
            String platform = args[ip + 1].toLowerCase();
            long accessCode = Long.parseLong(args[ic + 1]);
            if (accessCode < 1L || accessCode > 99999L)
            {
                System.out.println("Invalid access code. Valid access codes are in range 1 - 99999");
                return;
            }
            long[] array1;
            long[] array2;
            switch (platform)
            {
                case "pc":
                {
                        array1 = new long[] { 331L, 1259L, 241L, 109L, 521L, 853L, 71L, 719L, 941L, 269L };
                        array2 = new long[] { 72481L, 180307L, 130241L, 392827L, 421019L, 949147L, 32801L, 1296649L, 91249L, 639679L };
                        break;
                }

                case "ps2":
                case "xbox":
                {
                        array1 = new long[] { 859L, 773L, 151L, 47L, 487L, 211L, 617L, 131L, 947L, 313L };
                        array2 = new long[] { 69119L, 67783L, 70271L, 77929L, 238099L, 148151L, 472751L, 818963L, 1195489L, 839381L };
                        break;
                }

                case "psp":
                {
                        array1 = new long[] { 743L, 1663L, 227L, 991L, 443L, 89L, 571L, 199L, 1373L, 601L };
                        array2 = new long[] { 35491L, 783019L, 1116491L, 591319L, 194591L, 37369L, 822839L, 86083L, 354661L, 99809L };
                        break;
                }

                default:
                {
                        System.out.println("Invalid platform. Enter a platform among pc, ps2, psp, xbox");
                        return;
                }
            }
            System.out.println("Platform: " + platform.toUpperCase());
            System.out.println("Access code: " + accessCode);
            for (int i = 0; i < cheats.length; i++)
            {
                System.out.println(cheats[i] + ": " + GenerateCode(array1, array2, accessCode, i));
            }
        }

        private static String GenerateCode(long[] array1, long[] array2, long accessCode, long cheatID)
        {
            long cheatIDMagic = 0x13CB5BL * cheatID % 0x26DDL;
            long accessCodeMagic = (accessCode % 0x3E8L) ^ (0x20B9L * cheatIDMagic);
            long seed1 = CalculateSeed(accessCodeMagic % 0x9ADL);
            long seed2 = CalculateSeed(Remainder(ToSigned32((accessCodeMagic ^ 0x114CF1L) * ((0x41BL * cheatIDMagic) ^ Remainder(Quotient(accessCode, 0x3E8L), 0x3E8L))), 0x91DL));
            long[] buffer = { 0L, 0L, 0L, 0L, 0L, 0L };
            buffer[0] = Remainder(seed1, 26L) + 65L;
            buffer[1] = Remainder(Quotient(seed1, 676L), 26L) + 65L;
            buffer[2] = Remainder(Quotient(seed1, 26L), 26L) + 65L;
            buffer[3] = Remainder(Quotient(seed2, 26L), 26L) + 65L;
            buffer[4] = Remainder(Quotient(seed2, 676L), 26L) + 65L;
            buffer[5] = Remainder(seed2, 26L) + 65L;
            long bufMidXor = CalculateFeedback(buffer);
            long feedback1 = ToSigned32((buffer[0] << 24) + (buffer[1] << 16) + (buffer[2] << 8) + buffer[3]);
            long feedback2 = ToSigned32((buffer[4] << 24) + (buffer[5] << 16) + ((bufMidXor + Remainder(cheatIDMagic ^ 0x197ABD9L, seed1 & 0xFFFFFFFFL)) << 8) + bufMidXor + Remainder(cheatIDMagic ^ 0x13478FDDL, seed2 & 0xFFFFFFFFL));
            long tmpFeedback2 = 0L;
            for (int i = 0; i < 42; i++)
            {
                tmpFeedback2 = feedback2;
                feedback2 = feedback1 ^ array1[i % 10];
                feedback1 = tmpFeedback2 ^ feedback1;
            }
            for (int i = 0; i < 277; i++)
            {
                tmpFeedback2 = feedback2;
                feedback2 = feedback1 ^ array2[i % 10];
                feedback1 = feedback1 ^ tmpFeedback2;
            }
            buffer[0] = Remainder((feedback2 >> 24) & 0xFFL, 26L) + 65L;
            buffer[1] = Remainder((feedback2 >> 16) & 0xFFL, 26L) + 65L;
            buffer[2] = Remainder((feedback1 >> 24) & 0xFFL, 26L) + 65L;
            buffer[3] = Remainder((feedback1 >> 16) & 0xFFL, 26L) + 65L;
            buffer[4] = Remainder((feedback1 >> 8) & 0xFFL, 26L) + 65L;
            buffer[5] = Remainder(feedback1 & 0xFFL, 26L) + 65L;
            String code = "";
            for (int i = 0; i < 6; i++)
            {
                code += (char) buffer[i];
            }
            return code;
        }

        private static long Quotient(long n, long m)
        {
            return n / m;
        }

        private static long Remainder(long n, long m)
        {
            return n % m;
        }

        private static long CalculateFeedback(long[] buffer)
        {
            long result = 0;
            for (int i = 0; i < buffer.length - 1; i++)
            {
                result += buffer[i] ^ 0x13C501L;
            }
            return ToSigned32(result);
        }

        private static long ToSigned32(long n)
        {
            return ((n & 0xFFFFFFFFL) ^ 0x80000000L) - 0x80000000L;
        }

        private static long CalculateSeed(long n)
        {
            long seed = 1L;
            if (n != 0L)
            {
                seed = 0xF82DL;
                for (int i = 0; i < n - 1; i++)
                {
                    seed = Remainder(ToSigned32(0xF82DL * seed), 0x5243L);
                }
            }
            return seed;
        }
    }
